docs/proposals/nodeselector.md

```diff
@@ -0,0 +1,261 @@
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NodeSelector and NodeAffinity
+
+## Introduction
+
+This document proposes a new label selector representation, called `NodeSelector`,
+that is similar in many ways to `LabelSelector`, but is a bit more flexible and is
+intended to be used only for selecting nodes.
+
+In addition, we propose to replace the `map[string]string` in `PodSpec` that the scheduler
+currently uses as part of restricting the set of nodes onto which a pod is
+eligible to schedule, with a field of type `Affinity` that contains contains one or
+more affinity specifications. In this document we discuss `NodeAffinity`, which
+contains one or more of the following
+* a field called `RequiredDuringSchedulingRequiredDuringExecution` that will be
+represented by a `NodeSelector`, and thus generalizes the scheduling behavior of
+the current `map[string]string` but still serves the purpose of restricting
+the set of nodes onto which the pod can schedule. In addition, unlike the behavior
+of the current `map[string]string`, when it becomes violated the system will
+try to eventually evict the pod from its node.
+* a field called `RequiredDuringSchedulingIgnoredDuringExecution` which is identical
+to `RequiredDuringSchedulingRequiredDuringExecution` except that the system
+may or may not try to eventually evict the pod from its node.
+* a field called `PreferredDuringSchedulingIgnoredDuringExecution` that specifies which nodes are
+preferred for scheduling among those that meet all scheduling requirements.
+
+(In practice, as discussed later, we will actually *add* the `Affinity` field
+rather than replacing `map[string]string`, due to backward compatibility requirements.)
+
+The affiniy specifications described above allow a pod to request various properties
+that are inherent to nodes, for example "run this pod on a node with an Intel CPU" or, in a
+multi-zone cluster, "run this pod on a node in zone Z."
+([This issue](https://github.com/kubernetes/kubernetes/issues/9044) describes
+some of the properties that a node might publish as labels, which affinity expressions
+can match against.)
+They do *not* allow a pod to request to schedule
+(or not schedule) on a node based on what other pods are running on the node. That
+feature is called "inter-pod topological affinity/anti-afinity" and is described
+[here](https://github.com/kubernetes/kubernetes/pull/18265).
+
+## API
+
+### NodeSelector
+
+```go
+// A node selector represents the union of the results of one or more label queries
+// over a set of nodes; that is, it represents the OR of the selectors represented
+// by the nodeSelectorTerms.
+type NodeSelector struct {
+	// nodeSelectorTerms is a list of node selector terms. The terms are ORed.
+	NodeSelectorTerms []NodeSelectorTerm `json:"nodeSelectorTerms,omitempty"`
+}
+
+// An empty node selector term matches all objects. A null node selector term
+// matches no objects.
```

Comment:
Since we updated the semantic, there's no different matching behaviors between empty and nil in the sub-level of NodeSelector:
>
1. nil NodeSelector matches all nodes (i.e. does not filter out any nodes)
2. nil []NodeSelectorTerm (equivalent to non-nil empty NodeSelector) matches no nodes
3. zero-length non-nil []NodeSelectorTerm we can say matches no nodes also, just for simplicity
4. nil []NodeSelectorRequirement (equivalent to non-nil empty NodeSelectorTerm) matches no nodes
5. zero-length non-nil []NodeSelectorRequirement we can say matches no nodes also, just for simplicity
6. non-nil empty NodeSelectorRequirement is not allowed (this is already implemented by virtue of using LabelSelectorRequirement to implement NodeSelectorRequirement): https://github.com/kubernetes/kubernetes/blob/725a84ffe02d854fdafe14e4c86718736cb20da7/pkg/apis/extensions/helpers.go#L56 )

And since `NodeSelectorTerms` is tagged with `omitempty`, non-nil zero []NodeSelectorTerm, will be regarded as nil in json.Marshal()